include/shadesmar/memory/dragons.h

```diff
@@ -196,12 +205,158 @@ class AvxUnrollCopier : public Copier {
+}
+
+class AvxAsyncUnrollCopier : public Copier {
+ public:
+  using PtrT = __m256i;
+  constexpr static size_t alignment = 4 * sizeof(__m256i);
+
+  void *alloc(size_t size) override {
+    return aligned_alloc(alignment, SHMALIGN(size, alignment));
+  }
+
+  void dealloc(void *ptr) override { free(ptr); }
+
+  void shm_to_user(void *dst, void *src, size_t size) override {
+    _avx_async_cpy_unroll(dst, src, SHMALIGN(size, alignment));
+  }
+
+  void user_to_shm(void *dst, void *src, size_t size) override {
+    _avx_async_cpy_unroll(dst, src, SHMALIGN(size, alignment));
+  }
+};
+
+//------------------------------------------------------------------------------
+
+static inline void _avx_async_pf_cpy_unroll(void *d, const void *s, size_t n) {
+  // d, s -> 128 byte aligned
+  // n -> multiple of 128
+
+  auto *dVec = reinterpret_cast<__m256i *>(d);
+  const auto *sVec = reinterpret_cast<const __m256i *>(s);
+  size_t nVec = n / sizeof(__m256i);
+  for (; nVec > 4; nVec -= 4, sVec += 4, dVec += 4) {
+    // prefetch data for next iteration
+    _mm_prefetch(sVec + 4, _MM_HINT_T0);
+    _mm_prefetch(sVec + 6, _MM_HINT_T0);
+    _mm256_stream_si256(dVec, _mm256_load_si256(sVec));
+    _mm256_stream_si256(dVec + 1, _mm256_load_si256(sVec + 1));
+    _mm256_stream_si256(dVec + 2, _mm256_load_si256(sVec + 2));
+    _mm256_stream_si256(dVec + 3, _mm256_load_si256(sVec + 3));
+  }
+  _mm256_stream_si256(dVec, _mm256_load_si256(sVec));
+  _mm256_stream_si256(dVec + 1, _mm256_load_si256(sVec + 1));
+  _mm256_stream_si256(dVec + 2, _mm256_load_si256(sVec + 2));
+  _mm256_stream_si256(dVec + 3, _mm256_load_si256(sVec + 3));
+  _mm_sfence();
+}
+
+class AvxAsyncPFUnrollCopier : public Copier {
+ public:
+  using PtrT = __m256i;
+  constexpr static size_t alignment = 4 * sizeof(__m256i);
+
+  void *alloc(size_t size) override {
+    return aligned_alloc(alignment, SHMALIGN(size, alignment));
+  }
+
+  void dealloc(void *ptr) override { free(ptr); }
+
+  void shm_to_user(void *dst, void *src, size_t size) override {
+    _avx_async_pf_cpy_unroll(dst, src, SHMALIGN(size, alignment));
+  }
+
+  void user_to_shm(void *dst, void *src, size_t size) override {
+    _avx_async_pf_cpy_unroll(dst, src, SHMALIGN(size, alignment));
+  }
+};
+
+//------------------------------------------------------------------------------
+
+template <class BaseCopierT, uint32_t nthreads>
+class MTCopier : public Copier {
+ public:
+  MTCopier() : base_copier() {}
+
+  void *alloc(size_t size) override { return base_copier.alloc(size); }
+
+  void dealloc(void *ptr) override { base_copier.dealloc(ptr); }
+
+  void _copy(void *d, void *s, size_t n, bool shm_to_user) {
+    n = SHMALIGN(n, sizeof(typename BaseCopierT::PtrT)) /
+        sizeof(typename BaseCopierT::PtrT);
+    std::vector<std::thread> threads;
+    threads.reserve(nthreads);
+
+#ifdef __APPLE__
+    lldiv_t per_worker = div((int64_t)n, nthreads);
+#else
+    ldiv_t per_worker = div((int64_t)n, nthreads);
+#endif
```

Comment:
Gotta love this stupid difference.

Reply:
Won't this be solvable if [std::lldiv_t](https://en.cppreference.com/w/cpp/numeric/math/div) is explicitly called?

Reply:
Brilliant; good catch.

Reply:
Better yet: `auto per_worker = ...`

Reply:
Hahaha yeah, that's the best way.